lib/matplotlib/category.py

```diff
@@ -29,18 +29,23 @@ def convert(value, unit, axis):
         Parameters
         ----------
         value : string or iterable
-            value or list of values to be converted
-        unit : :class:`.UnitData`
-           object string unit information for value
-        axis : :class:`~matplotlib.Axis.axis`
-            axis on which the converted value is plotted
+            Value or list of values to be converted.
+        unit : `.UnitData`
+            An object mapping strings to integers.
+        axis : `~matplotlib.axis.Axis`
+            axis on which the converted value is plotted.
+
+            .. note:: *axis* is unused.
 
         Returns
         -------
-        mapped_ value : float or ndarray[float]
-
-        .. note:: axis is not used in this function
+        mapped_value : float or ndarray[float]
         """
+        if unit is None:
+            raise ValueError(
+                'Missing unit for StrCategoryConverter. This might be caused'
```

Comment:
So what's missing here is the mapping from category to float. Worried about using unit in a user facing error because unit is a kwarg that doesn't have to be passed in for categoricals to work.

Reply:
> So what's missing here is the mapping from category to float. 

Not sure what that means. Moving the example from pyplot to ax to make it a little more obvious:

~~~
ax = plt.gca()
ax.bar([1, 2, 3], [50, 10, 20])
ax.set_xticks(['banana', 'apple', 'pear'])
~~~

We are trying to string positions (not labels) to a float axis. IMHO there's nothing missing here and this should error out. Ideally, the error message would be something like "Attemtping to set categorical ticks to a non-categorical axis".

But there's no good place to raise that.

The original stacktrace before this PR was:
~~~
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-13c713392566> in <module>()
      2 ax = plt.gca()
      3 ax.bar([1, 2, 3], [50, 10, 20])
----> 4 ax.set_xticks(['banana', 'apple', 'pear'])

~/dev/matplotlib/lib/matplotlib/axes/_base.py in set_xticks(self, ticks, minor)
   3303             Default is ``False``.
   3304         """
-> 3305         ret = self.xaxis.set_ticks(ticks, minor=minor)
   3306         self.stale = True
   3307         return ret

~/dev/matplotlib/lib/matplotlib/axis.py in set_ticks(self, ticks, minor)
   1686         """
   1687         # XXX if the user changes units, the information will be lost here
-> 1688         ticks = self.convert_units(ticks)
   1689         if len(ticks) > 1:
   1690             xleft, xright = self.get_view_interval()

~/dev/matplotlib/lib/matplotlib/axis.py in convert_units(self, x)
   1488             return x
   1489 
-> 1490         ret = self.converter.convert(x, self.units, self)
   1491         return ret
   1492 

~/dev/matplotlib/lib/matplotlib/category.py in convert(value, unit, axis)
     51 
     52         # force an update so it also does type checking
---> 53         unit.update(values)
     54 
     55         str2idx = np.vectorize(unit._mapping.__getitem__,

AttributeError: 'NoneType' object has no attribute 'update'
~~~

So, you can raise
1. in set_ticks(self, ticks, minor)
   *Advantage:* You know you are setting ticks, so you could issue "Attemtping to set categorical ticks to a non-categorical axis".
   *Disadvantage:* You need to explicitly check ticks and the unit of Axis.
2. in convert_units(self, x)
   *Advantage:* You can investigate self.units and generate a more appropriate message related to categoricals.
   *Disadvantage:* You don't know that you are setting ticks here. You would have to explicitly check which converter you are using.
3. in convert(value, unit, axis)
   *Advantage:* You know that unit must not be *None* so the check is trivial.
   *Disadvantage:* You don't know anything about categorical axes anymore and therefore can only state that unit is missing and add a guess why that might be the case.

In that situation I decided to raise on 3. add add context information in 2. using the `raise .. from`.

Reply:
I'm not against the error, I worry that the specific word `unit` will lead users down a google rabbit hole that's misguided. Basically, there's an xunit/yunit kwarg:
https://matplotlib.org/gallery/units/bar_demo2.html#sphx-glr-gallery-units-bar-demo2-py
```python
axs[1, 0].bar(cms, cms, bottom=bottom, width=width, xunits=inch, yunits=cm)
```
So I worry an error message with the word `unit` might make users think the fix is to pass in something, when really the issue is that the axis input data is not a string so the axis is not registered as string units. I wonder if the following is enough: 
```
'Axis does not support StrCategoryConvertor; axis data values must be strings'
```
Like you, I want to help people catch that the problem is the mismatch between the plot data and the tick data.

Reply:
Rephrased to

~~~
            raise ValueError(
                'Missing category information for StrCategoryConverter; '
                'this might be caused by unintendedly mixing categorical and '
                'numeric data')
~~~

I don't want to mention an Axis as we don't know the call context here.